exported default function might break AOT

<!--
PLEASE HELP US PROCESS GITHUB ISSUES FASTER BY PROVIDING THE FOLLOWING INFORMATION.

ISSUES MISSING IMPORTANT INFORMATION MIGHT BE CLOSED WITHOUT INVESTIGATION.
-->

## I'm submitting a ...
<!-- Check one of the following options with "x" -->
<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[X] Bug report <!-- Please search github for a similar issue or PR before submitting -->
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Below code fails with:
```
ERROR in Cannot read property 'text' of undefined

ERROR in ./src/main.ts
Module not found: Error: Can't resolve './$$_gendir/app/app.module.ngfactory' in 'HOME/demo-app/src'
 @ ./src/main.ts 4:0-74
 @ multi ./src/main.ts
```

## Expected behavior
That code:
```
export default function () {

  const mainRoutes = [
    {name: 'a', abstract: true, component: 'main'},

    {name: 'a.welcome', url: '/welcome', component: 'welcome'}
  ];

  return mainRoutes;

}
```
should compile fine under AOT. it's actually a simple array. removing the unncessary variable, should work, too.

## Minimal reproduction of the problem with instructions
Create a file called `states.ts` somewhere in your source folder (does not need to be imported) and insert the code above.
Run `ng build --prod`.


## What is the motivation / use case for changing the behavior?

## Please tell us about your environment

Angular version: 4.2.2

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: v8.1.1
- Platform: Mac OS X 11.11

Others:
Might be related, but different error message: https://github.com/angular/angular/issues/11402
Looks like that works:
```
export default [
    {name: 'a', abstract: true, component: 'main'},

    {name: 'a.welcome', url: '/welcome', component: 'welcome'}
];
```
which however is also default

--- a/tools/@angular/tsc-wrapped/src/collector.ts
+++ b/tools/@angular/tsc-wrapped/src/collector.ts
@@ -377,7 +377,7 @@ export class MetadataCollector {
           // Record functions that return a single value. Record the parameter
           // names substitution will be performed by the StaticReflector.
           const functionDeclaration = <ts.FunctionDeclaration>node;
-          if (isExported(functionDeclaration)) {
+          if (isExported(functionDeclaration) && functionDeclaration.name) {
             if (!metadata) metadata = {};
             const name = exportedName(functionDeclaration);
             const maybeFunc = maybeGetSimpleFunction(functionDeclaration);

--- a/tools/@angular/tsc-wrapped/test/collector.spec.ts
+++ b/tools/@angular/tsc-wrapped/test/collector.spec.ts
@@ -634,10 +634,6 @@ describe('Collector', () => {
   });
 
   describe('with interpolations', () => {
-    function createSource(text: string): ts.SourceFile {
-      return ts.createSourceFile('', text, ts.ScriptTarget.Latest, true);
-    }
-
     function e(expr: string, prefix?: string) {
       const source = createSource(`${prefix || ''} export let value = ${expr};`);
       const metadata = collector.getMetadata(source);
@@ -821,16 +817,52 @@ describe('Collector', () => {
 
   describe('regerssion', () => {
     it('should be able to collect a short-hand property value', () => {
-      const source = ts.createSourceFile(
-          '', `
+      const source = createSource(`
         const children = { f1: 1 };
         export const r = [
           {path: ':locale', children}
         ];
-      `,
-          ts.ScriptTarget.Latest, true);
+      `);
       const metadata = collector.getMetadata(source);
       expect(metadata.metadata).toEqual({r: [{path: ':locale', children: {f1: 1}}]});
+    });
+
+    // #17518
+    it('should skip a default function', () => {
+      const source = createSource(`
+        export default function () {
+
+          const mainRoutes = [
+            {name: 'a', abstract: true, component: 'main'},
+
+            {name: 'a.welcome', url: '/welcome', component: 'welcome'}
+          ];
+
+          return mainRoutes;
+
+        }`);
+      const metadata = collector.getMetadata(source);
+      expect(metadata).toBeUndefined();
+    });
+
+    it('should skip a named default export', () => {
+      const source = createSource(`
+        function mainRoutes() {
+
+          const mainRoutes = [
+            {name: 'a', abstract: true, component: 'main'},
+
+            {name: 'a.welcome', url: '/welcome', component: 'welcome'}
+          ];
+
+          return mainRoutes;
+
+        }
+
+        exports = foo;
+        `);
+      const metadata = collector.getMetadata(source);
+      expect(metadata).toBeUndefined();
     });
   });
 
@@ -1321,3 +1353,7 @@ const FILES: Directory = {
     }
   }
 };
+
+function createSource(text: string): ts.SourceFile {
+  return ts.createSourceFile('', text, ts.ScriptTarget.Latest, true);
+}

